Steadily decrease veto power as PR ages

pavellishin opened the pull request:
As the PR gets older, require more and more rejections to "veto" it, overriding any non-unanimous majority approvals.

Thus, a newly opened PR can be blocked with a single veto (0 days + 1).

A PR that has been open for 3 days requires four rejections to be vetoed (3 days + 1).

I believe this will help PRs from being stalled forever by rejections, while addressing @tnelling's [concerns](https://github.com/jeffkaufman/nomic/pull/29#pullrequestreview-189647685) about an overly-fast paced game.

- [ ] Waiting on #37 to get merged to more accurately gauge a PR's age.

tnelling commented:
I like this idea.

I'd like to make it a little more intuitive, and have the rejection count needed be n, rather than n+1.

This has the side effect that at <1 day it's (arguably) blocked with 0 rejections (no matter how many approvals). I could see keeping it at 1 for <1 day as a special case. I could also see treating that side effect as a way to give everyone time to make double check things without having them rammed through in a matter of hours or even minutes of coordinated action, because you always have a day to review.*

If others disagree, I'll accept this as-is.

*I realize that at present you always have a day to review anyway, but this would be a way to codify that notion independent of the approval threshold decay logic, which will be particularly relevant if we go to 2/3 majority per #31.

jeffkaufman commented on validate.py:
```diff
@@ -117,6 +118,12 @@ def determine_if_mergeable():
+  # overriding any non-unanimous majority approvals.
+  # Thus, a newly opened PR can be blocked with a single veto (0 days + 1)
+  # A PR that has been open for 3 days requires four rejections to be vetoed.
+  rejections_required_to_veto = math.floor(days_since_last_commit() + 1)
```
`days_since_last_commit` isn't a good way to measure PR age.  I can submit a new PR with an old commit on it, and it will read as old.  The functionality you need is in #37

jeffkaufman commented:
@pavellishin you'll probably need to wait until #37 is merged

jeffkaufman commented:
As written this PR decreases the value of a veto based on how many days it's been since something was last merged to master.  Which isn't what it was intended to do, ok, but is also a fine thing to do.

jeffkaufman commented:
@pavellishin #37 is in now if you want to pick this back up

tnelling commented on validate.py:
```diff
@@ -1,3 +1,4 @@
+import math
```
extra import of math

pavellishin replied:
makes you smarter

tnelling commented on validate.py:
```diff
@@ -207,6 +208,15 @@ def determine_if_mergeable():
+  # A PR that has been open for 3 days requires four rejections to be vetoed.
+  # (Note that this is the number of days since the PR was opened, not since its last change.)
+  pr_creation_timestamp = pr_created_at_ts(pr_json)
+  pr_age_in_days = math.floor(seconds_to_days(seconds_since(pr_creation_timestamp)))
```
There are utility functions this can leverage

pavellishin replied:
hm? Do you mean you'd like to see me write some, or that some exist? I didn't see anything like `days_since`.

jeffkaufman replied:
The `math.floor` isn't needed, since `seconds_to_days` already truncates.

tnelling replied:
I might be wrong about what exactly is in there, or getting confused with an unmerged PR. If you don't see anything, go ahead and resolve this. I had the impression you hadn't re-evaluated since your initial commit, but I may be mistaken.

pavellishin replied:
I probably need to give it another sober pass, I had like 10 minutes of downtime at work in which I did this. I need to test it locally.

jeffkaufman commented on validate.py:
```diff
@@ -217,8 +226,14 @@ def determine_if_mergeable():
+    raise Exception('Veto: rejected by: %s of required %s for %s-day-old PR (%s)' % (
+        len(rejections),
+        rejections_required_to_veto,
+        math.floor(days_since_last_commit()),
```
`math.floor` not needed here either

jeffkaufman commented on validate.py:
```diff
@@ -207,6 +207,15 @@ def determine_if_mergeable():
+  # (Note that this is the number of days since the PR was opened, not since its last change.)
+  pr_creation_timestamp = pr_created_at_ts(pr_json)
+  pr_age_in_days = math.floor(seconds_to_days(seconds_since(pr_creation_timestamp)))
+  rejections_required_to_veto = math.floor(pr_age_in_days + 1)
```
`math.floor` isn't needed here, since pr_age_in_days was already an integer

jeffkaufman commented:
I'll approve once these are fixed

pavellishin commented:
Ok, I've updated (and also fixed) the outstanding issues.
 - renamed some functions for clarity
 - actual bugfix
 - no more redundant flooring
 - more output

@jeffkaufman @tnelling 
and also @shaunagm @sligocki @thejhyde @vesche

jeffkaufman commented on validate.py:
```diff
@@ -167,18 +167,21 @@ def last_commit_ts():
 
+def days_since(ts):
+  return seconds_to_days(seconds_since(ts))
+
```
duplicate of a function below

pavellishin replied:
Whoops, I did a bad job rebasing

jeffkaufman commented on validate.py:
```diff
@@ -213,6 +213,19 @@ def determine_if_mergeable():
+  # Thus, a newly opened PR can be blocked with a single veto (0 days + 1)
+  # A PR that has been open for 3 days requires four rejections to be vetoed.
+  # (Note that this is the number of days since the PR was opened, not since its last change.)
+  pr_creation_timestamp = pr_created_at_ts(pr_json)
```
Wait, this should be `pr_last_changed_ts`

jeffkaufman replied:
What's the point of vetoing then?  Why not remove the requirement entirely?

pavellishin replied:
Two things:

1. It allows a player to basically demand time to do a more thorough inspection of a PR. If I see something shifty in the morning before I have to run to work, I can reject a PR so I can take a look at it in the evening, even if the rest of the group approves it.

2. It allows a player to basically buy time to argue against a PR; it's sort of analogous to a filibuster. If everyone decides a PR is 👍, but I feel very strongly about not merging it in, I can block it for at least some amount of time while I make my case.

jeffkaufman replied:
In the case of a single rejection (now that #44 is merged) what this PR does is just get rid of rejection blocking.  You already can't merge a PR within a day of its creation without 100% agreement.  So neither of your examples make sense to me.

Instead, I think you should just write comments.  No one wants bad PRs merged, so just write things like "let's take our time reviewing this PR, there's risky stuff in it" or "I think this PR has a subtle flaw around timing, but I'm not going to have time to explain it until tomorrow -- please hold off on approving it for now".

pavellishin commented:
But fair enough, your comments make sense; I'll close this PR and re-open it if we have issues in the future.
